{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'\\nIn Python, exceptions are used to handle errors and exceptional conditions that may occur during the execution of a program. The Exception class is the base class for all built-in exceptions in Python, and it provides a common interface for handling exceptions.\\n\\nWhen creating a custom exception, we need to inherit from the Exception class (or one of its subclasses) to ensure that our custom exception can be caught and handled using the standard exception handling mechanisms in Python. This allows us to define our own exception types that can be used to signal specific error conditions or exceptional situations in our code.'"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Q1. Explain why we have to use the Exception class while creating a Custom Exception.\n",
    "#Note: Here Exception class refers to the base class for all the exceptions.\n",
    "'''\n",
    "In Python, exceptions are used to handle errors and exceptional conditions that may occur during the execution of a program. The Exception class is the base class for all built-in exceptions in Python, and it provides a common interface for handling exceptions.\n",
    "\n",
    "When creating a custom exception, we need to inherit from the Exception class (or one of its subclasses) to ensure that our custom exception can be caught and handled using the standard exception handling mechanisms in Python. This allows us to define our own exception types that can be used to signal specific error conditions or exceptional situations in our code.'''\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "pwskills input.\n"
     ]
    }
   ],
   "source": [
    "\n",
    "# Suggestion: Add a code example of creating a custom exception\n",
    "class CustomException(Exception):\n",
    "    \"\"\"A custom exception for signaling invalid input.\"\"\"\n",
    "\n",
    "    def __init__(self, message=\"pwskills input.\"):\n",
    "        \"\"\"Initialize the custom exception with a message.\"\"\"\n",
    "        self.message = message\n",
    "        super().__init__(self.message)\n",
    "\n",
    "# Example usage:\n",
    "try:\n",
    "    raise CustomException()\n",
    "except CustomException as e:\n",
    "    print(e)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Python Exception Hierarchy:\n",
      "TypeError\n",
      "  FloatOperation\n",
      "  MultipartConversionError\n",
      "StopAsyncIteration\n",
      "StopIteration\n",
      "ImportError\n",
      "  ModuleNotFoundError\n",
      "    PackageNotFoundError\n",
      "  ZipImportError\n",
      "OSError\n",
      "  ConnectionError\n",
      "    BrokenPipeError\n",
      "    ConnectionAbortedError\n",
      "    ConnectionRefusedError\n",
      "    ConnectionResetError\n",
      "      RemoteDisconnected\n",
      "  BlockingIOError\n",
      "  ChildProcessError\n",
      "  FileExistsError\n",
      "  FileNotFoundError\n",
      "  IsADirectoryError\n",
      "  NotADirectoryError\n",
      "  InterruptedError\n",
      "    InterruptedSystemCall\n",
      "  PermissionError\n",
      "  ProcessLookupError\n",
      "  TimeoutError\n",
      "  UnsupportedOperation\n",
      "  itimer_error\n",
      "  herror\n",
      "  gaierror\n",
      "  SSLError\n",
      "    SSLCertVerificationError\n",
      "    SSLZeroReturnError\n",
      "    SSLWantWriteError\n",
      "    SSLWantReadError\n",
      "    SSLSyscallError\n",
      "    SSLEOFError\n",
      "  Error\n",
      "    SameFileError\n",
      "  SpecialFileError\n",
      "  ExecError\n",
      "  ReadError\n",
      "  URLError\n",
      "    HTTPError\n",
      "    ContentTooShortError\n",
      "  BadGzipFile\n",
      "EOFError\n",
      "  IncompleteReadError\n",
      "RuntimeError\n",
      "  RecursionError\n",
      "  NotImplementedError\n",
      "    ZMQVersionError\n",
      "    StdinNotImplementedError\n",
      "  _DeadlockError\n",
      "  BrokenBarrierError\n",
      "  BrokenExecutor\n",
      "    BrokenThreadPool\n",
      "  SendfileNotAvailableError\n",
      "  VariableError\n",
      "NameError\n",
      "  UnboundLocalError\n",
      "AttributeError\n",
      "  FrozenInstanceError\n",
      "SyntaxError\n",
      "  IndentationError\n",
      "    TabError\n",
      "LookupError\n",
      "  IndexError\n",
      "  KeyError\n",
      "    NoSuchKernel\n",
      "    UnknownBackend\n",
      "  CodecRegistryError\n",
      "ValueError\n",
      "  UnicodeError\n",
      "    UnicodeEncodeError\n",
      "    UnicodeDecodeError\n",
      "    UnicodeTranslateError\n",
      "  UnsupportedOperation\n",
      "  JSONDecodeError\n",
      "  SSLCertVerificationError\n",
      "  Error\n",
      "  UnsupportedDigestmodError\n",
      "  IllegalMonthError\n",
      "  IllegalWeekdayError\n",
      "  ParserError\n",
      "  MessageDefect\n",
      "    NoBoundaryInMultipartDefect\n",
      "    StartBoundaryNotFoundDefect\n",
      "    CloseBoundaryNotFoundDefect\n",
      "    FirstHeaderLineIsContinuationDefect\n",
      "    MisplacedEnvelopeHeaderDefect\n",
      "    MissingHeaderBodySeparatorDefect\n",
      "    MultipartInvariantViolationDefect\n",
      "    InvalidMultipartContentTransferEncodingDefect\n",
      "    UndecodableBytesDefect\n",
      "    InvalidBase64PaddingDefect\n",
      "    InvalidBase64CharactersDefect\n",
      "    InvalidBase64LengthDefect\n",
      "    HeaderDefect\n",
      "      InvalidHeaderDefect\n",
      "      HeaderMissingRequiredValue\n",
      "      NonPrintableDefect\n",
      "      ObsoleteHeaderDefect\n",
      "      NonASCIILocalPartDefect\n",
      "      InvalidDateDefect\n",
      "  ClassNotFound\n",
      "  ClipboardEmpty\n",
      "  MacroToEdit\n",
      "  InvalidVersion\n",
      "AssertionError\n",
      "ArithmeticError\n",
      "  FloatingPointError\n",
      "  OverflowError\n",
      "  ZeroDivisionError\n",
      "    DivisionByZero\n",
      "    DivisionUndefined\n",
      "  DecimalException\n",
      "    Clamped\n",
      "    Rounded\n",
      "      Underflow\n",
      "      Overflow\n",
      "    Inexact\n",
      "      Underflow\n",
      "      Overflow\n",
      "    Subnormal\n",
      "      Underflow\n",
      "    DivisionByZero\n",
      "    FloatOperation\n",
      "    InvalidOperation\n",
      "      ConversionSyntax\n",
      "      DivisionImpossible\n",
      "      DivisionUndefined\n",
      "      InvalidContext\n",
      "SystemError\n",
      "  CodecRegistryError\n",
      "ReferenceError\n",
      "MemoryError\n",
      "BufferError\n",
      "Warning\n",
      "  UserWarning\n",
      "    GetPassWarning\n",
      "    FormatterWarning\n",
      "  EncodingWarning\n",
      "  DeprecationWarning\n",
      "    ProvisionalWarning\n",
      "  PendingDeprecationWarning\n",
      "  SyntaxWarning\n",
      "  RuntimeWarning\n",
      "    ProactorSelectorThreadWarning\n",
      "    UnknownTimezoneWarning\n",
      "  FutureWarning\n",
      "    ProvisionalCompleterWarning\n",
      "  ImportWarning\n",
      "  UnicodeWarning\n",
      "  BytesWarning\n",
      "  ResourceWarning\n",
      "  DeprecatedTzFormatWarning\n",
      "_OptionError\n",
      "_Error\n",
      "error\n",
      "Verbose\n",
      "Error\n",
      "SubprocessError\n",
      "  CalledProcessError\n",
      "  TimeoutExpired\n",
      "ArgumentError\n",
      "error\n",
      "ZMQBaseError\n",
      "  ZMQError\n",
      "    ContextTerminated\n",
      "    Again\n",
      "    InterruptedSystemCall\n",
      "  ZMQBindError\n",
      "  NotDone\n",
      "Error\n",
      "PickleError\n",
      "  PicklingError\n",
      "  UnpicklingError\n",
      "_Stop\n",
      "TokenError\n",
      "StopTokenizing\n",
      "Error\n",
      "  CancelledError\n",
      "  TimeoutError\n",
      "  InvalidStateError\n",
      "_GiveupOnSendfile\n",
      "Incomplete\n",
      "ClassFoundException\n",
      "EndOfBlock\n",
      "TimeoutError\n",
      "InvalidStateError\n",
      "LimitOverrunError\n",
      "QueueEmpty\n",
      "QueueFull\n",
      "TraitError\n",
      "Empty\n",
      "Full\n",
      "error\n",
      "error\n",
      "ReturnValueIgnoredError\n",
      "ArgumentError\n",
      "ArgumentTypeError\n",
      "ConfigError\n",
      "  ConfigLoaderError\n",
      "    ArgumentError\n",
      "  ConfigFileNotFound\n",
      "ConfigurableError\n",
      "  MultipleInstanceError\n",
      "ApplicationError\n",
      "InvalidPortNumber\n",
      "error\n",
      "LZMAError\n",
      "RegistryError\n",
      "_GiveupOnFastCopy\n",
      "Error\n",
      "  NoSectionError\n",
      "  DuplicateSectionError\n",
      "  DuplicateOptionError\n",
      "  NoOptionError\n",
      "  InterpolationError\n",
      "    InterpolationMissingOptionError\n",
      "    InterpolationSyntaxError\n",
      "    InterpolationDepthError\n",
      "  ParsingError\n",
      "    MissingSectionHeaderError\n",
      "NoIPAddresses\n",
      "Error\n",
      "BadZipFile\n",
      "LargeZipFile\n",
      "Error\n",
      "MessageError\n",
      "  MessageParseError\n",
      "    HeaderParseError\n",
      "    BoundaryError\n",
      "  MultipartConversionError\n",
      "  CharsetError\n",
      "DuplicateKernelError\n",
      "ErrorDuringImport\n",
      "NotOneValueFound\n",
      "CannotEval\n",
      "OptionError\n",
      "BdbQuit\n",
      "Restart\n",
      "FindCmdError\n",
      "HomeDirError\n",
      "ProfileDirError\n",
      "IPythonCoreError\n",
      "  TryNext\n",
      "  UsageError\n",
      "  StdinNotImplementedError\n",
      "InputRejected\n",
      "GetoptError\n",
      "ErrorToken\n",
      "PrefilterError\n",
      "AliasError\n",
      "  InvalidAliasError\n",
      "Error\n",
      "  InterfaceError\n",
      "  DatabaseError\n",
      "    InternalError\n",
      "    OperationalError\n",
      "    ProgrammingError\n",
      "    IntegrityError\n",
      "    DataError\n",
      "    NotSupportedError\n",
      "Warning\n",
      "ExceptionGroup\n",
      "SpaceInInput\n",
      "DOMException\n",
      "  IndexSizeErr\n",
      "  DomstringSizeErr\n",
      "  HierarchyRequestErr\n",
      "  WrongDocumentErr\n",
      "  InvalidCharacterErr\n",
      "  NoDataAllowedErr\n",
      "  NoModificationAllowedErr\n",
      "  NotFoundErr\n",
      "  NotSupportedErr\n",
      "  InuseAttributeErr\n",
      "  InvalidStateErr\n",
      "  SyntaxErr\n",
      "  InvalidModificationErr\n",
      "  NamespaceErr\n",
      "  InvalidAccessErr\n",
      "  ValidationErr\n",
      "ValidationError\n",
      "EditReadOnlyBuffer\n",
      "_Retry\n",
      "InvalidLayoutError\n",
      "HeightIsUnknownError\n",
      "GuardRejection\n",
      "ParserSyntaxError\n",
      "InternalParseError\n",
      "_PositionUpdatingFinished\n",
      "SimpleGetItemNotFound\n",
      "UncaughtAttributeError\n",
      "HasNoContext\n",
      "ParamIssue\n",
      "_JediError\n",
      "  InternalError\n",
      "  WrongVersion\n",
      "  RefactoringError\n",
      "OnErrorLeaf\n",
      "InvalidPythonEnvironment\n",
      "HTTPException\n",
      "  NotConnected\n",
      "  InvalidURL\n",
      "  UnknownProtocol\n",
      "  UnknownTransferEncoding\n",
      "  UnimplementedFileMode\n",
      "  IncompleteRead\n",
      "  ImproperConnectionState\n",
      "    CannotSendRequest\n",
      "    CannotSendHeader\n",
      "    ResponseNotReady\n",
      "  BadStatusLine\n",
      "    RemoteDisconnected\n",
      "  LineTooLong\n",
      "InteractivelyDefined\n",
      "KillEmbedded\n",
      "Error\n",
      "  NoSuchProcess\n",
      "    ZombieProcess\n",
      "  AccessDenied\n",
      "  TimeoutExpired\n",
      "_Ipv6UnsupportedError\n",
      "KeyReuseError\n",
      "UnknownKeyError\n",
      "LeakedCallbackError\n",
      "BadYieldError\n",
      "ReturnValueIgnoredError\n",
      "Return\n",
      "QueueEmpty\n",
      "QueueFull\n",
      "DebuggerInitializationError\n",
      "ExpatError\n",
      "Error\n",
      "  ProtocolError\n",
      "  ResponseError\n",
      "  Fault\n",
      "_Error\n",
      "UnableToResolveVariableException\n",
      "InvalidTypeInArgsException\n",
      "CustomException\n",
      "CustomException\n",
      "CustomException\n",
      "CustomException\n"
     ]
    }
   ],
   "source": [
    "import pprint\n",
    "\n",
    "def print_exception_hierarchy(exc, level=0):\n",
    "    \"\"\"Print the exception hierarchy starting from the given exception class.\n",
    "\n",
    "    Args:\n",
    "    exc (type): The exception class to start the hierarchy from.\n",
    "    level (int): The current level of indentation.\n",
    "    \"\"\"\n",
    "    for sub_exc in exc.__subclasses__():\n",
    "        print(\"  \" * level + sub_exc.__name__)\n",
    "        print_exception_hierarchy(sub_exc, level + 1)\n",
    "\n",
    "if __name__ == \"__main__\":\n",
    "    print(\"Python Exception Hierarchy:\")\n",
    "    print_exception_hierarchy(Exception)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "division by zero\n"
     ]
    }
   ],
   "source": [
    "#Q3. What errors are defined in the ArithmeticError class? Explain any two with an example.\n",
    "'''The ArithmeticError class is a built-in Python exception class that serves as the base class for arithmetic errors. '''\n",
    "#Here are two examples of arithmetic errors:\n",
    "\n",
    "#1.ZeroDivisionError:\n",
    "try:\n",
    "    result = 1 / 0\n",
    "except ZeroDivisionError as e:\n",
    "    print(e)\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "#2.OverflowError:\n",
    "\n",
    "try:\n",
    "    result = 1e300 * 1e300\n",
    "except OverflowError as e:\n",
    "    print(e)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [
    {
     "ename": "KeyError",
     "evalue": "'address'",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mKeyError\u001b[0m                                  Traceback (most recent call last)",
      "Cell \u001b[0;32mIn[10], line 10\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[38;5;124;03m'''\u001b[39;00m\n\u001b[1;32m      4\u001b[0m \u001b[38;5;124;03mThe LookupError class in Python is the base class for exceptions that are raised when a key or index is not found in a collection. It has two main subclasses: KeyError and IndexError.\u001b[39;00m\n\u001b[1;32m      5\u001b[0m \n\u001b[1;32m      6\u001b[0m \u001b[38;5;124;03mA KeyError is raised when a key is not found in a dictionary. For example:\u001b[39;00m\n\u001b[1;32m      7\u001b[0m \n\u001b[1;32m      8\u001b[0m \u001b[38;5;124;03m'''\u001b[39;00m\n\u001b[1;32m      9\u001b[0m my_dict \u001b[38;5;241m=\u001b[39m {\u001b[38;5;124m'\u001b[39m\u001b[38;5;124mname\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mJohn\u001b[39m\u001b[38;5;124m'\u001b[39m, \u001b[38;5;124m'\u001b[39m\u001b[38;5;124mage\u001b[39m\u001b[38;5;124m'\u001b[39m: \u001b[38;5;241m30\u001b[39m}\n\u001b[0;32m---> 10\u001b[0m \u001b[38;5;28mprint\u001b[39m(\u001b[43mmy_dict\u001b[49m\u001b[43m[\u001b[49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[38;5;124;43maddress\u001b[39;49m\u001b[38;5;124;43m'\u001b[39;49m\u001b[43m]\u001b[49m)  \u001b[38;5;66;03m# Raises a KeyError\u001b[39;00m\n",
      "\u001b[0;31mKeyError\u001b[0m: 'address'"
     ]
    }
   ],
   "source": [
    "#Q4. Why LookupError class is used? Explain with an example KeyError and IndexError.\n",
    "\n",
    "'''\n",
    "The LookupError class in Python is the base class for exceptions that are raised when a key or index is not found in a collection. It has two main subclasses: KeyError and IndexError.\n",
    "\n",
    "A KeyError is raised when a key is not found in a dictionary. For example:\n",
    "\n",
    "'''\n",
    "my_dict = {'name': 'John', 'age': 30}\n",
    "print(my_dict['address'])  # Raises a KeyError\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'         \\nImportError is a built-in exception in Python that occurs when an import statement fails to find the module, function, or object that you are trying to import. One common subclass of `ImportError\\n\\nModuleNotFoundError is a subclass of ImportError in Python. It is raised when a specified module cannot be found. This error typically occurs when you try to import a module that is not installed in your Python environment.\\n\\nTo avoid ModuleNotFoundError, make sure that the module you want to import is installed in your Python environment. You can use pip, the Python package installer, to install modules. For example, to install the module requests, you can run the following command in your terminal:\\n\\n\\n\\n\\n'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Q5. Explain ImportError. What is ModuleNotFoundError?\n",
    "\n",
    "'''         \n",
    "ImportError is a built-in exception in Python that occurs when an import statement fails to find the module, function, or object that you are trying to import. One common subclass of `ImportError\n",
    "\n",
    "ModuleNotFoundError is a subclass of ImportError in Python. It is raised when a specified module cannot be found. This error typically occurs when you try to import a module that is not installed in your Python environment.\n",
    "\n",
    "To avoid ModuleNotFoundError, make sure that the module you want to import is installed in your Python environment. You can use pip, the Python package installer, to install modules. For example, to install the module requests, you can run the following command in your terminal:\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Q6. List down some best practices for exception handling in python.\n",
    "\n",
    "'''Sure, here are some best practices for exception handling in Python:\n",
    "\n",
    "Use specific exceptions: Instead of using the generic Exception class, use more specific exceptions that are more informative and allow you to handle different types of errors in different ways.\n",
    "Don't catch all exceptions: Catching all exceptions can make it difficult to identify and debug errors. Instead, catch only the exceptions that you expect and know how to handle.\n",
    "Use finally to clean up resources: Use the finally clause to ensure that any resources that were opened or allocated are properly closed or released, even if an exception is raised.\n",
    "Use try-except-else: Use the else clause to execute code that should only run if no exceptions were raised. This can help to simplify your code and avoid unnecessary checks.\n",
    "Don't ignore exceptions: Don't just catch exceptions and then ignore them. Instead, log the exception or display an error message to the user.\n",
    "Raise exceptions for unexpected conditions: If a function encounters an unexpected condition, raise an exception to signal that something went wrong. This can help to make your code more robust and easier to debug.\n",
    "Use context managers: Use context managers to ensure that resources are properly acquired and released. This can help to simplify your code and avoid errors.\n",
    "Don't use exceptions for flow control: Don't use exceptions for flow control, such as using an exception to exit a loop. Instead, use more appropriate control structures like break or return.\n",
    "Provide meaningful error messages: When raising or catching exceptions, provide meaningful error messages that describe the problem and provide context for debugging.\n",
    "Test your exception handling code: Make sure to test your exception handling code thoroughly, including edge cases and unexpected conditions. This can help to ensure that your code handles errors gracefully and recovers from them effectively.\n",
    "\n",
    "\n",
    "\n",
    "\n",
    "'''\n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.10.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
